### PR TITLE
Catalog & Core Improvements

### PR DESCRIPTION
**Pull request by wearpants**
Infrastructure improvements `core` & fixes for `catalog`.

This adds a `core.models.FilesModel` that knows how to build a dataframe from input files. It's backed by https://github.com/mbr/simplekv, and works on both local files & S3, with transparent support for gzip files. 

Simplekv supports a key-value model, not a directory-of-files, so I wasn't able to port the existing code easily. `FilesModel` has a hook for listing the files, so it'd be easy to pass a pre-built list of filenames (I'm doing this with Crossref; `list_warcs()` job output could work here too).

This should make development much easier - I'd appreciate if we can merge soon & avoid another long fork.

### CODE REVIEW

File: osp_pipeline/core/models.py [hunk 1]
```diff
@@ -56,3 +63,65 @@ def from_rdd(cls, row):
...
+
+        if path.endswith(".gz"):
+            f = gzip.GzipFile(fileobj=f)
+
+        return f
+
+    @classmethod
+    def from_files(cls, url):
+        """build a dataframe from input files at `url`"""
+        spark = SparkSession.builder.getOrCreate()
```
Comment (davidmcclure) [inline osp_pipeline/core/models.py]: I realize I'm not sure how that `getOrCreate()` method works - will this pick up on the `spark` instance that gets created in the test suite?

File: osp_pipeline/core/models.py [hunk 2]
```diff
@@ -56,3 +63,65 @@ def from_rdd(cls, row):
...
+               .map(lambda f: partial(cls._open_file, kv, f))
+               .flatMap(cls.iterfile)
+               )
+
+        return spark.createDataFrame(rdd, cls.schema)
+
+    @classmethod
+    def get_filenames(cls, kv):
+        """return a RDD of file paths"""
+        sc = SparkContext.getOrCreate()
```
Comment (davidmcclure) [inline osp_pipeline/core/models.py]: Likewise here - do we need to pass this in, like we do with the job functions themselves?

**Comment by davidmcclure** [comment]
Hey @wearpants,

This looks great - I love the idea of the `FilesModel`. The only question I had was whether it works, from a testing standpoint, to directly call those `getOrCreate()` methods to get `spark` and `sc` instances, or if we have to somehow pass them from without, like we do with the job functions. (Or if it does work to just do `getOrCreate()`, should we do this with the job functions too?)

Before merging - looks like there's some test breakage.

```
_____________________________________________________________________ ERROR collecting  ______________________________________________________________________
env/lib/python3.6/site-packages/_pytest/config.py:337: in _importconftest
    mod = conftestpath.pyimport()
env/lib/python3.6/site-packages/py/_path/local.py:650: in pyimport
    __import__(modname)
env/lib/python3.6/site-packages/_pytest/assertion/rewrite.py:207: in load_module
    py.builtin.exec_(co, mod.__dict__)
tests/osp_pipeline/institutions/conftest.py:6: in <module>
    from osp_pipeline.institutions.models import Institution
osp_pipeline/institutions/models.py:16: in <module>
    from ..core.models import Model, FilesystemModel
E   ImportError: cannot import name 'FilesystemModel'

During handling of the above exception, another exception occurred:
env/lib/python3.6/site-packages/py/_path/common.py:332: in visit
    for x in Visitor(fil, rec, ignore, bf, sort).gen(self):
env/lib/python3.6/site-packages/py/_path/common.py:378: in gen
    for p in self.gen(subdir):
env/lib/python3.6/site-packages/py/_path/common.py:367: in gen
    dirs = self.optsort([p for p in entries
env/lib/python3.6/site-packages/py/_path/common.py:368: in <listcomp>
    if p.check(dir=1) and (rec is None or rec(p))])
env/lib/python3.6/site-packages/_pytest/main.py:672: in _recurse
    ihook = self.gethookproxy(path)
env/lib/python3.6/site-packages/_pytest/main.py:575: in gethookproxy
    my_conftestmodules = pm._getconftestmodules(fspath)
env/lib/python3.6/site-packages/_pytest/config.py:314: in _getconftestmodules
    mod = self._importconftest(conftestpath)
env/lib/python3.6/site-packages/_pytest/config.py:339: in _importconftest
    raise ConftestImportFailure(conftestpath, sys.exc_info())
E   _pytest.config.ConftestImportFailure: ImportError("cannot import name 'FilesystemModel'",)
E     File "/opt/osp/env/lib/python3.6/site-packages/_pytest/assertion/rewrite.py", line 207, in load_module
E       py.builtin.exec_(co, mod.__dict__)
E     File "/opt/osp/tests/osp_pipeline/institutions/conftest.py", line 6, in <module>
E       from osp_pipeline.institutions.models import Institution
E     File "/opt/osp/osp_pipeline/institutions/models.py", line 16, in <module>
E       from ..core.models import Model, FilesystemModel
```

I tried getting rid of that `FilesystemModel` import in the institution models, which got the suite to run, but with errors in the institutions tests and a failure in VIAF.

Thanks for this!

File: osp_pipeline/core/models.py (hunk 1 above)
Comment (wearpants) [inline osp_pipeline/core/models.py]: Sure will!

File: osp_pipeline/core/models.py (hunk 2 above)
Comment (wearpants) [inline osp_pipeline/core/models.py]: Nope, it's a global.

Comment (wearpants) [inline osp_pipeline/core/models.py]: I used this in Institutions too

**Pull request merged**